### PR TITLE
[MIN-4792] Change the format of exclude list to support http methods

### PR DESCRIPTION
**Pull request by zackxu1**
[MIN-4792 Change exclude list format](https://mindfulchef.atlassian.net/browse/MIN-4792)

## Changes

- Change the exclude list format to have `path` and `methods` (which can be either `:all` or a select list like `[:get]`)
- If a request path is among the exclude list, we still parse and verify the JWT token if one is supplied.

### CODE REVIEW

File: lib/rack/jwt/auth.rb [hunk 1]
```diff
@@ -64,25 +64,30 @@ def initialize(app, opts = {})
...
-        end
-        if auth_cookie_enabled? && empty_auth_cookie?(env)
-          return return_error('Empty token cookie')
-        end
-        if auth_cookie_disabled? && missing_auth_header?(env)
-          return return_error('Missing Authorization header')
-        end
-        if auth_cookie_disabled? && invalid_auth_header?(env)
-          return return_error('Invalid Authorization header format')
+        if auth_is_required
```
Comment (mattgibson) [inline lib/rack/jwt/auth.rb:69]:
```suggestion
        if auth_is_required?
```

File: lib/rack/jwt/auth.rb [hunk 2]
```diff
@@ -159,20 +164,30 @@ def check_valid_algorithm!
       end
 
       def check_exclude_type!
-        raise ArgumentError, 'exclude argument must be an Array' unless @exclude.is_a?(Array)
+        raise ArgumentError, 'exclude argument must be an Array' unless exclude.is_a?(Array)
 
-        @exclude.each do |x|
-          raise ArgumentError, 'each exclude Array element must be a String' unless x.is_a?(String)
-          raise ArgumentError, 'each exclude Array element must not be empty' if x.empty?
+        exclude.each do |x|
```
Comment (mattgibson) [inline lib/rack/jwt/auth.rb]:
```suggestion
        exclude.each do |exclusion|
```

File: lib/rack/jwt/auth.rb (hunk 1 above)
Comment (zackxu1) [reply lib/rack/jwt/auth.rb:69]: this is a local variable, i can't use a trailing `?` in the name.  i can turn it into a function, but I will have to do the exclude list lookup twice.

File: lib/rack/jwt/auth.rb (hunk 2 above)
Comment (zackxu1) [reply lib/rack/jwt/auth.rb]: Have updated the name.

File: lib/rack/jwt/auth.rb (hunk 1 above)
Comment (mattgibson) [reply lib/rack/jwt/auth.rb:69]: Ah, I hadn't notced it was a variable. Fair enough, then.

**Review by mattgibson — APPROVED** [review]

File: spec/token_spec.rb [hunk 3]
```diff
@@ -344,4 +344,57 @@
...
+          header 'Authorization', "Bearer #{issuer.encode(payload, nil, 'none')}"
+          get('/book')
+          expect(last_response.status).to eq 200
+          expect(last_response.headers['jwt.header']).to eq({"typ"=>"JWT", "alg"=>"none"})
+          expect(last_response.headers['jwt.payload']).to eq("foo" => "bar")
+        end
+      end
+
+      describe 'when JWT token is missing from header' do
+        it 'returns a 401' do
```
Comment (zorbash) [inline spec/token_spec.rb]: aggregate_failures

File: spec/auth_spec.rb [hunk 4]
```diff
@@ -116,31 +116,45 @@
...
+          args = { secret: secret, exclude: [ { path: '/static', foo: :all } ] }
           expect { Rack::JWT::Auth.new(inner_app, args) }.to raise_error(ArgumentError)
         end
       end
 
-      describe 'when Array contains elements that do not start with a /' do
+      describe 'when Array contains element whose path does not start with a /' do
         it 'raises an exception' do
-          args = { secret: secret, exclude: ['/foo', 'bar', '/baz'] }
+          args = { secret: secret, exclude: [ { path: 'static', methods: :all } ] }
```
Comment (zorbash) [inline spec/auth_spec.rb]: [nit] Consider removing whitespace before and after `{` and `}`

File: lib/rack/jwt/auth.rb [hunk 5]
```diff
@@ -159,20 +164,30 @@ def check_valid_algorithm!
...
+            raise ArgumentError, 'each exclude Array element must have only path and methods as keys'
+          end
 
-          unless x.start_with?('/')
-            raise ArgumentError, 'each exclude Array element must start with a /'
+          unless exclusion[:path].start_with?('/')
+            raise ArgumentError, 'each exclude Array element path value must start with a /'
+          end
+
+          unless exclusion[:methods] == :all || exclusion[:methods].is_a?(Array)
```
Comment (zorbash) [inline lib/rack/jwt/auth.rb]: This isn't backwards compatible. If the `methods` is key is omitted it'll raise an exception.

**Review by zorbash — CHANGES_REQUESTED** [review]
I think there's a backwards compatibility issue. When `methods` is not set it raises an error so it can be versioned as 0.7.0.

**Comment by wilsonsilva** [comment]
Would be cool if we could get eparreno to merge these changes in the gem.

**Comment by zackxu1** [comment]
@zorbash I have made the changes to make the exclude list backward compatible.  So each exclude element can be either a String or a Hash (in which case keys must be present: `:path` and `:methods`)

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,12 @@
 # Change Log
 
-## [0.6.0](https://github.com/eparreno/rack-jwt/tree/v0.5.0) (2020-09-29)
+## [0.7.0](https://github.com/eparreno/rack-jwt/tree/v0.7.0) (2021-04-27)
```
Comment (zorbash) [inline CHANGELOG.md]: [nit] Don't forget to update the date here

File: spec/auth_spec.rb [hunk 7]
```diff
@@ -116,31 +116,105 @@
...
+          expect { Rack::JWT::Auth.new(inner_app, args) }.not_to raise_error
+        end
+      end
+
+      describe 'when exclude list contains an array of valid hashes' do
+        it 'does not raise any exception' do
+          args = {
+            secret: secret,
+            exclude: [
+              { path: '/books', methods: :all },
```
Comment (zorbash) [inline spec/auth_spec.rb:132]: If one sets `methods: [:all]` will it work?

File: spec/auth_spec.rb [hunk 8]
```diff
@@ -116,31 +116,105 @@
...
+      describe 'when Array contains empty string' do
+        it 'raises an exception', :aggregate_failures do
+          args = { secret: secret, exclude: ['/foo', ''] }
+          expect { Rack::JWT::Auth.new(inner_app, args) }.to raise_error(ArgumentError)
+        end
+      end
+
+      describe 'when Array element contains a string not starting with /' do
+        it 'raises an exception', :aggregate_failures do
+          args = { secret: secret, exclude: %w(/boks music) }
```
Comment (zorbash) [inline spec/auth_spec.rb]: [nit]

`boks` here looks like a typo.

File: CHANGELOG.md (hunk 6 above)
Comment (zackxu1) [reply CHANGELOG.md]: 👍 Date updated

File: spec/auth_spec.rb (hunk 7 above)
Comment (zackxu1) [reply spec/auth_spec.rb:132]: I have made the change to allow this to work.  So if you specify `[:all]`, it's equivalent to `:all`

**Review by zorbash — APPROVED** [review]

**Pull request merged**